### DISCUSSION TITLE
Outlet velocity under fully developed flow using INS AD

### DISCUSSION THREAD

**Discussion by smpark7** [idaholab/moose#24834] [Q&A Modules: Navier-Stokes] [answered]
The INS FE AD system produces velocity upshoots/downshoots at the outlet for pipe/channel flow under fully developed flow conditions.

This can be reproduced using the [pipe flow test](https://github.com/idaholab/moose/blob/next/modules/navier_stokes/test/tests/finite_element/ins/rz-x-axial-coord/pipe-flow.i) by refining the mesh (e.g., nx=800, ny=20), lengthening the pipe from 1m to 4m, and reducing the Reynolds number to 100 (for faster transition to fully developed flow).

The figure shows the x velocity from x=3m to 4m along y=0m. An upshoot appears near the outlet after a region of relatively stable fully developed flow.
<img width="469" alt="image" src="https://github.com/idaholab/moose/assets/30237775/e5e385a3-d865-41d7-96d6-da845823c308">

Refining the mesh near the outlet boundary does not help and may actually worsen the solution where the flow transitions from coarse to fine mesh regions, possibly due to reduced SUPG stabilization from the smaller mesh sizes.

I can't figure out the numerical reasoning behind this behavior. Is there any way to suppress the upshoots/downshoots?

**Comment by lindsayad** [comment]
I'm not seeing it. I applied this patch
```
diff --git a/modules/navier_stokes/test/tests/finite_element/ins/rz-x-axial-coord/pipe-flow.i b/modules/navier_stokes/test/tests/finite_element/ins/rz-x-axial-coord/pipe-flow.i
index a609bf2107..58de9a989a 100644
--- a/modules/navier_stokes/test/tests/finite_element/ins/rz-x-axial-coord/pipe-flow.i
+++ b/modules/navier_stokes/test/tests/finite_element/ins/rz-x-axial-coord/pipe-flow.i
@@ -1,8 +1,8 @@
 # 2d siumulation of a water through a pipe.
 mu=1e-3 # Nsm^-2
 rho=997.0 # kgm^-3
-Re=1000.0
-pipe_length=1 # m
+Re=100.0
+pipe_length=4 # m
 pipe_radius=0.1 # m
 u_inlet=${fparse (mu * Re)/(2 * pipe_radius * rho)} # ms^-1
 
@@ -18,8 +18,8 @@ u_inlet=${fparse (mu * Re)/(2 * pipe_radius * rho)} # ms^-1
     xmax = ${pipe_length}
     ymin = 0
     ymax = ${pipe_radius}
-    nx = 50
-    ny = 5
+    nx = 800
+    ny = 20
   []
 []
 
```
and the radial velocity profiles look identical to me at x = pipe-length/2 and x = pipe-length
<img width="665" alt="Screen Shot 2023-06-27 at 12 00 06 PM" src="https://github.com/idaholab/moose/assets/10248304/95908d7b-c560-410a-8ec5-d5cdbc456563">

**Reply by lindsayad** [reply]
Oh ok I do see what you see. Is this really a big concern though? I mean this is the plot with a y-axis from 0 to .002
<img width="662" alt="Screen Shot 2023-06-27 at 12 03 55 PM" src="https://github.com/idaholab/moose/assets/10248304/3f9fe86f-e4e3-42c6-95d5-d02eab0042dc">

**Reply by lindsayad** [reply, marked as answer]
I tried both Dirichlet and natural conditions at the outlet and the results are pretty much identical

<img width="1182" alt="Screen Shot 2023-06-27 at 12 25 21 PM" src="https://github.com/idaholab/moose/assets/10248304/abba052e-f9a4-48d8-a882-6e79392773c6">

**Reply by smpark7** [reply]
My main concern is mass/energy conservation under this type of flow over a long period of time. Does this fall within expectations for non-conservative behavior in FEM flow modeling?

**Reply by lindsayad** [reply]
For discussion on conservativeness of continuous finite elements, I refer you to https://www.sciencedirect.com/science/article/abs/pii/S002199910096577X which includes discussion about Dirichlet boundary conditions.

You may also explore turning off stabilization terms as those introduce volumetric `_grad_test` objects which implicitly add their own flux contribution at boundaries, e.g. a natural BC for momentum will zero out the sum of stresses, pressure (if the latter is integrated by parts), *and* stabilization

**Reply by lindsayad** [reply]
BTW investigating this I did find some bugs that I'm fixing in https://github.com/idaholab/moose/pull/24842

**Reply by smpark7** [reply]
Thanks for the reference and the stabilization suggestion. I'll look into them and report my findings here if they're interesting.

**Reply by smpark7** [reply]
Just to clarify, what exactly do you mean by "Dirichlet BC at the outlet"? Do you mean fixing the x velocity to a parabolic profile?

**Reply by lindsayad** [reply]
No I believe that pipe flow test uses a Dirichlet BC for pressure at the outlet as it doesn't integrate the pressure by parts

**Reply by smpark7** [reply]
For the natural BC case, I assume you applied integration by parts? Did you write your own BC object for the pressure contribution along the axial boundary?

**Reply by lindsayad** [reply]
> For the natural BC case, I assume you applied integration by parts?

yes

> Did you write your own BC object for the pressure contribution along the axial boundary?

No. We are already setting the y-velocity in a Dirichlet way there. And there is no pressure contribution to the x-velocity equation along the axis-boundary since the normal is in the -y direction

**Comment by smpark7** [comment]
@lindsayad I noticed that the `INSADMomentumViscous` kernel overrides the `computeResidual` method and it does not check for displaced meshes. Could it be the reason for the discrepancies between INS and INSAD results in #24042?

Also, I was wondering what's the difference between accumulating residuals using `accumulateTaggedLocalResidual` and `_assembly.processResiduals`.

**Reply by lindsayad** [reply]
> I noticed that the `INSADMomentumViscous` kernel overrides the `computeResidual` method and it does not check for displaced meshes. Could it be the reason for the discrepancies between INS and INSAD results in #24042?

There is no need to check for displacements in the residual method. Data members like `_JxW` always contains the right numbers; it just doesn't contain any derivative information for Jacobian computation. That's why we do the checks in the Jacobian computation. @eiloo was getting different answers because they were not setting `use_displaced_mesh = true` in the INSAD materials. 

> Also, I was wondering what's the difference between accumulating residuals using accumulateTaggedLocalResidual and _assembly.processResiduals

There is no difference in terms of results. (`_assembly.processResiduals` has actually been removed)

**Reply by smpark7** [reply]
Okay thanks for clarifying!
